pandas/core/arrays/period.py

```diff
@@ -121,18 +122,30 @@ def freq(self, value):
 
     _attributes = ["freq"]
 
-    def __new__(cls, values, freq=None, **kwargs):
+    def __new__(cls, values, freq=None, dtype=None, **kwargs):
+
+        if freq is not None:
+            # coerce freq to freq object, otherwise it can be coerced
+            # elementwise, which is slow
+            freq = Period._maybe_convert_freq(freq)
+
+        freq = dtl.validate_dtype_freq(dtype, freq)
+
         if is_period_dtype(values):
             # PeriodArray, PeriodIndex
-            if freq is not None and values.freq != freq:
-                raise IncompatibleFrequency(freq, values.freq)
-            freq = values.freq
+            freq = dtl.validate_dtype_freq(values.dtype, freq)
             values = values.asi8
 
         elif is_datetime64_dtype(values):
-            # TODO: what if it has tz?
             values = dt64arr_to_periodarr(values, freq)
 
+        elif is_object_dtype(values) or isinstance(values, (list, tuple)):
```

Comment:
shouldn't this be is_list_like? (for the isinstance check)

Reply:
This is specifically for object dtype (actually, I need to add `dtype=object` to the `np.array` call below) since we're calling libperiod.extract_ordinals, which expects object dtype.

Reply:
Other question: where was this handled previously?

Reply:
>Other question: where was this handled previously?

This was and still is handled in `PeriodIndex.__new__` in the catch-all block at the end of the method.  Since the PeriodArray constructor will not accept all the kwargs that the PeriodIndex constructor does, it can be appreciably simpler.

Reply:
I still don't fully understand why you are adding it. 
The question is maybe, should the PeriodArray constructor actually handle this? 
Or what other change in this PR necessitates adding it?

Reply:
> I still don't fully understand why you are adding it.

Because we want the following to work (and tests are added for it):

```
per = PeriodArray([Period('2016Q1'), Period('2016Q2')], freq='Q')
per2 = PeriodArray(np.array([Period('2016Q1'), Period('2016Q2')]), freq='Q')
```

Reply:
But do we? I am personally not sure we should do that. 
I raised the same question on @TomAugspurger's PR (which actually also indicates this PR is partly duplicating work that Tom is doing there)

Reply:
I think it would be very weird to not support this.  Basic round-trips should work:  `arr == PeriodArray(list(arr))` and `arr == PeriodArray(arr.astype(object))`

+1 on soliciting opinions

Reply:
BTW, I think we should also make this an `__init__` instead of `__new__`. 

Which is what Tom already did in his PR.

> I raised the same question on @TomAugspurger's PR (which actually also indicates this PR is partly duplicating work that Tom is doing there)

Further on this, I really have the feeling that we are duplicating both clean-up effort and discussion between you and what Tom is doing on the PeriodArray PR

Reply:
> I think we should also make this an __init__ instead of __new__.

I don't have a strong opinion on this at the moment.  General thought is that this is in the category of things to handle "at the end".

>  Further on this, I really have the feeling that we are duplicating both clean-up effort and discussion between you and what Tom is doing on the PeriodArray PR

Definitely on the conversation.  On the effort there is overlap but less duplication.  (this is why I tried to rebase the PeriodArray PR, suggested doing the PeriodDType parts separately)

Reply:
@TomAugspurger @jreback thoughts on handling object-dtype in PeriodArray constructor?

Reply:
I'd prefer that go through a dedicated constructor.

Reply:
> I'd prefer that go through a dedicated constructor.

I can see having that be the Suggested Usage, and fine if we want the base constructor to dispatch to it, but I'm still puzzled as to why we wouldn't want the main constructor to round-trip.

If jreback is with you guys on this I'll chalk this up as one of those things that I don't understand but have to accept (like the popularity of The Big Bang Theory, or the Compactness Theorem).

Reply:
It's hard for me to say what's better in the abstract.

From the WIP PeriodArray PR, I found that having to think carefully about what type of data I had forced some clarity in the code. I liked having to explicitly reach for that `_from_periods` constructor.

Regardless, I think our two goals with the array constructors should be

1. Maximizing developer happiness (i.e. not users at the moment)
2. Making it easier to reuse code between Index & Array subclasses

If you think we're likely to end up in a situation where being able to pass an array of objects to the main `__init__` will make things easier, then by all means.

Reply:
i am a bit puzzled why you would handle lists and and ndarray differently (tom and joris); these are clearly doing the same thing and we have a very similar handling for list likes throughout pandas 

separating these is a non starter - even having a separate constructor is also verbally not very friendly. pandas does inference on the construction which is one of the big selling points. trying to change this, esp at the micro level is a huge mental disconnect.

Reply:
> i am a bit puzzled why you would handle lists and and ndarray differently (tom and joris)

I don't think we are.

But, my only argument was

> From the WIP PeriodArray PR, I found that having to think carefully about what type of data I had forced some clarity in the code. I liked having to explicitly reach for that _from_periods constructor.

If that's not persuasive then I'm not going to argue against handling them in the init.

Reply:
> having to think carefully

+1

> Maximizing developer happiness

+1

> Making it easier to reuse code

+1

> If you think we're likely to end up in a situation where being able to pass an array of objects to the main

Yes, I think we should be pretty forgiving about what gets accepted into `__init__` (for all three of Period/Datetime/Timedelta Arrays).  Definitely don't want the `start, end, periods` currently in the Index subclass constructors.  I think by excluding those we'll keep these constructors fairly straightforward.

Reply:
> i am a bit puzzled why you would handle lists and and ndarray differently 

It's not about lists vs arrays, it's about arrays of Period objects vs arrays of ordinal integers, which is something very different.

> I think we should be pretty forgiving about what gets accepted into __init__

Being forgiving is exactly what lead to the complex Period/DatetimeIndex constructors. I think we should not make the same choice for our Array classes. 
Of course it doesn't need to be that complex, as I think there are here two main usecases discussed: an array of scalar objects (eg Periods or Timestamps), or an array of the underlying storage type (eg datetime64 or ordinal integers).  

I personally also think it makes the code clearer to even separate those two concepts (basically what we also did with IntegerArray), but maybe let's open an issue to further discuss that instead of here in a hidden review comment thread? (i can only open one later today )